pkg/controller/apiextensions/definer/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	if err := r.client.Get(ctx, req.NamespacedName, definer); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We don't
+		// need to take any action in that case.
+		log.Debug(errGetInfraDef, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetInfraDef)
+	}
+
+	log = log.WithValues(
+		"uid", definer.GetUID(),
+		"version", definer.GetResourceVersion(),
+		"name", definer.GetName(),
+	)
+
+	crd, err := r.crd.Get(ctx, definer)
+	if resource.IgnoreNotFound(err) != nil {
+		log.Debug(errGetCRD, "error", err)
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, errGetCRD)
+	}
+	exists := !kerrors.IsNotFound(err)
+
+	// We make sure the CRD that we'll start a controller for is not marked for
+	// deletion so that when it does get deleted our controller and cache don't
+	// crash. Since we are the ones explicitly calling deletion of CRD after
+	// all instances are gone, it should be OK not start the controller.
+	if exists && !meta.WasDeleted(crd) {
+		// TODO(muvaf): controller and establishment checks could go into controller
+		// engine.
+
+		// There is one-to-one relationship between definer and CRD.
+		if !metav1.IsControlledBy(crd, definer) {
+			log.Debug(errNotControllerByDefiner)
+			definer.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.New(errNotControllerByDefiner)))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, definer), errUpdateInfraDefStatus)
+		}
+
+		// todo: shorter wait?
+
+		// It takes a while for api-server to be able to work with the new kind.
+		if !v1alpha1.IsEstablished(*crd) {
+			log.Debug(waitingCRDEstablish)
+			definer.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.New(waitingCRDEstablish)))
+			return reconcile.Result{RequeueAfter: 3 * time.Second}, errors.Wrap(r.client.Status().Update(ctx, definer), errUpdateInfraDefStatus)
+		}
+
+		// We know that CRD is ready and we are in control of it. So, we'll spin up
+		// an instance controller to reconcile it.
+		reconciler := composed.NewInstanceReconciler(definer.GetCRDName(), r.mgr, definer.GetCRDGroupVersionKind(), r.log, definer)
+		if err := r.ctrl.Start(definer.GetCRDName(), definer.GetCRDGroupVersionKind(), reconciler); err != nil {
+			log.Debug(errCannotStartController, "error", err)
+			definer.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.Wrap(err, errCannotStartController)))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, definer), errCannotStartController)
+		}
+		// TODO(muvaf): Consider using _Established_ condition for Definers
+		// similar to CRD.
+
+		// If the CRD is established and controller is up, then we are available.
+		definer.Status.SetConditions(runtimev1alpha1.Available())
+	}
+
+	if meta.WasDeleted(definer) {
```

Comment:
I'm guessing we can't just use Kubernetes GC to remove our CRD because we need to ensure we stop the controller first?

Reply:
Yes, we need to do a graceful ordered deletion like: delete instances, stop controller, delete CRD so that controller doesn't crash but it's worth a revisit later.